refactor ChainIndex types module into smaller modules

nachog00 opened the pull request:
## Motivation

The types.rs module on chain_index is too big (3333 lines). Splitting it into smaller, focused modules following the `module.rs + module/` pattern preferred at ZingoLabs.

Also added a new type for TreeSize that validates the size and safely handles edge cases, preventing dangerous `as u32` casts that could cause silent truncation when tree sizes reach their theoretical maximum.

## Solution

### Module Organization
- **`commitment.rs`**: Extracted commitment tree types (`CommitmentTreeData`, `CommitmentTreeRoots`, `CommitmentTreeSizes`) - ~160 lines
- **`primitives/`**: Created new primitives module structure:
  - `primitives.rs`: Index file for primitive types
  - `primitives/tree_size.rs`: Validated tree size types with bounds checking

### New Validated Tree Size Types
Added type-safe, validated representations of commitment tree sizes:
- **`TreeSize` trait**: Common behavior with validation, conversion methods, and bounds checking
- **`SproutTreeSize`**: Max 2^29 = 536,870,912 notes
- **`SaplingTreeSize`**: Max 2^32 = 4,294,967,296 notes  
- **`OrchardTreeSize`**: Max 2^32 = 4,294,967,296 notes
- **`TreeSizeError`**: Comprehensive error handling for edge cases

The TreeSize types address a critical issue where tree sizes are stored as u64 in Zebra but often cast to u32 in Zaino, creating potential for silent truncation when trees reach their theoretical maximum capacity (exactly u32::MAX + 1 for Sapling/Orchard).

### Tests

Compilation verified across all packages. TreeSize types include comprehensive validation and conversion methods with proper error handling.

### Specifications & References

Part of #538 - Replace custom block/transaction parsing with Zebra's consensus-validated deserialization

### Follow-up Work

The primitives module is prepared for migration of additional primitive types from the main types.rs file:
- Block and transaction hashes
- Block heights and chain work  
- Basic indexing primitives

### PR Checklist

- [x] The PR name is suitable for the release notes.
- [x] The solution is tested.
- [x] The documentation is up to date.

nachog00 committed:
Refactor: Move commitment tree types to dedicated module

Move CommitmentTreeData, CommitmentTreeRoots, and CommitmentTreeSizes
from the monolithic types.rs (3281 lines) to a dedicated commitment.rs
module. This is the first step in breaking down the large types file
into focused, manageable modules.

Changes:
- Create chain_index/types/commitment.rs with moved types
- Add module declaration and re-exports to types.rs
- Remove original type definitions from types.rs
- Maintain exact same functionality and API

Part of types modularization effort to improve code organization.

nachog00 committed:
Enhance: Add validated TreeSize types with trait-based design

Add semantically distinct, validated tree size types for each Zcash
shielded pool with a shared trait interface:

- TreeSize trait with default implementations for common operations
- SproutTreeSize (max 2^29 = 536,870,912 notes)
- SaplingTreeSize (max 2^32 = 4,294,967,296 notes)
- OrchardTreeSize (max 2^32 = 4,294,967,296 notes)
- TreeSizeError with descriptive error variants

Features:
- Compile-time pool-specific validation bounds
- Semantic type safety preventing pool confusion
- Safe u32 conversion with proper edge case handling
- Trait-based design eliminates code duplication
- Rich utility methods (is_at_max, remaining_capacity)
- Comprehensive documentation

This provides the foundation for eliminating dangerous 'as u32' casts
and prevents silent truncation errors when working with tree sizes.

nachog00 committed:
Refactor: Move TreeSize types to primitives module

Extract TreeSize trait and implementations from commitment.rs to a new
primitives module, following the module.rs + module/ pattern:

- Create types/primitives.rs as index module
- Create types/primitives/tree_size.rs with moved TreeSize types
- Update commitment.rs to focus only on commitment tree data structures
- Update re-exports in main types.rs

This establishes primitives/ as the home for foundational types that
other modules depend on, providing better separation of concerns and
preparing for extraction of other primitive types (BlockHash, Height, etc.).

Directory structure:
types/
├── primitives.rs          # Re-exports from primitives/
├── primitives/
│   └── tree_size.rs       # TreeSize trait + pool-specific implementations
└── commitment.rs          # Commitment tree data structures

nachog00 committed:
Fix: Update primitives module documentation

Correct the module documentation to accurately reflect current state:
- Currently contains only tree size types
- Document planned migrations from main types.rs
- Add migration strategy in git notes for reference

The previous documentation incorrectly claimed to contain types
that are still in the main types.rs file.

nachog00 committed:
Refactor: Reorganize types.rs with clear module structure

Restructure types.rs to clearly separate:

1. Module-level documentation explaining the transition strategy
2. Organized submodules (primitives, commitment) with re-exports
3. Legacy imports and types awaiting migration

This provides a clear roadmap for incremental migration and makes
the current state of the modularization effort explicit. The file
now serves as both a transition hub and a guide for future work.

Structure:
- MODULE ORGANIZATION: Clean submodule declarations
- IMPORTS FOR LEGACY TYPES: Support for un-migrated types
- LEGACY TYPES AWAITING MIGRATION: Clear migration candidates

nachog00 committed:
Merge branch 'dev' into refactor/modularize-types

nachog00 committed:
Merge branch 'dev' into refactor/modularize-types

nachog00 committed:
cargo fmt --all

nachog00 committed:
whitespace trip

nachog00 committed:
Fix broken intra-doc links in zaino-state types module

Updates documentation links to use proper module paths for primitives and commitment submodules.

nachog00 committed:
Merge branch 'dev' into refactor/modularize-types

nachog00 committed:
Merge branch 'dev' into refactor/modularize-types

dorianvp committed:
Merge branch 'dev' into refactor/modularize-types

idky137 commented on zaino-state/src/chain_index/types.rs:
```diff
@@ -1,6 +1,35 @@
+//! - `transaction.rs` - Transaction types (CompactTxData, TransparentCompactTx, etc.)
+//! - `address.rs` - Address and UTXO types (AddrScript, Outpoint, etc.)
+//! - `shielded.rs` - Shielded pool types (SaplingCompactTx, OrchardCompactTx, etc.)
+
```
Can we add a clear doc comment that outlines rules for this module:
- structs in this module _must_ _never_ use external types as fields directly, instead fundamental data should be saved into the struct, and from / into (or appropriate getters / setters) should be implemented.
- structs in this module _must_ implement `ZainoVersionedSerialize` _and_ abide by the stringent version rules outlined in that trait.
- structs in this module _must_ _never_ be changed without implementing a new version _and_ implementing the necessary ZainoDB updates and migrations.

AloeareV replied:
Agreed, but as these comments don't currently exist I don't think the need to add them should block this PR. I've opened #574 as follow-up work.

idky137 commented on zaino-state/src/chain_index/types/primitives/tree_size.rs:
```diff
@@ -0,0 +1,151 @@
+//! Validated commitment tree size types.
```
We currently have a CommitmentTreeSizes struct that holds the data that we require, changing this would require updates to ZainoDB (see comment above) i think the most straight forward way to gain the validation implemented here without wider cvhanges to the codebase would be to use the struct we already have and implement the new validation there.

AloeareV replied:
Currently these types aren't actually used anywhere. Any issues with how these types fit into the database won't come to light until we actually try to incorporate them, at which point they may prove to be incompatible with how our database is written/read. For that reason I'm hesitant to approve them. I'm a big fan of splitting `types.rs` into different modules for readability, and I don't want that to get blocked on discussion of new types.

idky137 reviewed (dismissed):
Separating out into sub-modules would be good but we need to be very careful with changing any of the structs in this file / module (see first comment).

fluidvanadium committed:
Merge branch 'dev' into refactor/modularize-types

nachog00 committed:
Merge branch 'dev' into refactor/modularize-types

nachog00 committed:
Remove tree size primitives from types
- Removed re-export of tree sizes in `types.rs`
- Deleted `tree_size.rs` and its contents from `primitives` module
- Updated `primitives.rs` to reflect migration plan for other types

Cleaned up code by removing unused tree size types and preparing for future refactoring to enhance module organization and clarity.

AloeareV reviewed (approved):
Moves only, this makes this clearer to read!
